src/main/java/eu/arrowhead/deviceqosevaluator/engine/SystemDeviceMap.java

```diff
@@ -0,0 +1,188 @@
+				devices.get(device).getRight().setValue(true);
+			}
+		}
+	}
+
+	//=================================================================================================
+	// assistant methods
+
+	//-------------------------------------------------------------------------------------------------
+	private Set<Address> collectAddresses(final SystemResponseDTO system) {
+		logger.debug("collectAddresses started");
+
+		final Set<Address> addresses = new HashSet<>(system.addresses().size());
+		for (final AddressDTO addr : system.addresses()) {
+			if (Utilities.isEnumValue(addr.type(), AddressType.class)) {
+				addresses.add(new Address(addr.address(), AddressType.valueOf(addr.type()), false));
+			}
+		}
+
+		if (system.device() != null) {
+			for (final AddressDTO addr : system.device().addresses()) {
+				if (Utilities.isEnumValue(addr.type(), AddressType.class)
+						&& !addr.type().equalsIgnoreCase(AddressType.MAC.name())) {
+					addresses.add(new Address(addr.address(), AddressType.valueOf(addr.type()), true));
+				}
+			}
+		}
+
+		return addresses;
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private Integer findDevice(final Set<Address> addresses) {
+		logger.debug("findDevice started");
+
+		for (final Entry<Integer, Triple<Set<Address>, Set<String>, Bool>> entry : devices.entrySet()) {
+			for (final Address deviceAddr : entry.getValue().getLeft()) {
+				final boolean anyMatch = addresses.stream().anyMatch(addr -> addr.address().equals(deviceAddr.address()));
+				if (anyMatch) {
+					return entry.getKey();
+				}
+			}
+		}
+
+		return null;
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	@SuppressWarnings("unchecked")
+	private boolean supportsAugmented(final SystemResponseDTO system) {
+		logger.debug("findDevice started");
+
+		if (Utilities.isEmpty(system.metadata()) || !system.metadata().containsKey(Constants.PROPERTY_KEY_QOS)) {
+			return false;
+		}
+
+		final Object object = MetadataKeyEvaluator.getMetadataValueForCompositeKey(system.metadata(), Constants.PROPERTY_KEY_QOS + Constants.DOT + Constants.PROPERTY_KEY_DEVICE_AUGMENTED);
+		if (object != null) {
+			try {
+				final List<String> list = (List<String>) object;
+				for (final String item : list) {
+					for (final OidGroup oidGroup : OidGroup.values()) {
+						if (item.equals(oidGroup.getValue())) {
+							return true;
+						}
+					}
+				}
+			} catch (final ClassCastException ex) {
+				return false;
+			}
+		}
+
+		return false;
+	}
+
+	//=================================================================================================
+	// nested record
+
+	protected record Address(String address, AddressType type, boolean deviceRelated) {
+	}
+
+	//=================================================================================================
+	// nested class
+	private final class Bool {
```

Comment:
why do you need custom boolean type?

Reply:
to ensure that the value is always a primitive boolean, nut also need an object to us it in the HashMap

Reply:
but the Bool object itself still can be null.